The Article Analyzer has been implemented.

volyamnyi opened the pull request.

volyamnyi committed:
The Article Analyzer has been implemented.

volyamnyi committed:
Merge pull request #1 from volyamnyi/yamnyi-article-analyzer

The Article Analyzer has been implemented.

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/ArticleAnalyzerApplication.java:
```diff
@@ -0,0 +1,22 @@
+
+        articleService.analyze();
+        articleService.saveCategoriesToJSONFileAsArray("categories");
+        articleService.notifyUsers(applicationContext);
```
move this logic to some class (maybe ArticleAnalyzer or you name it)

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/model/User.java:
```diff
@@ -0,0 +1,14 @@
+@Data
+public class User {
+    private Integer id;
+    @SerializedName("username")
```
nice

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,25 @@
+
+@Repository
+public class ArticleRepository {
+    @Autowired
```
use constructor injection instead of field one

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,25 @@
+public class ArticleRepository {
+    @Autowired
+    JsonManagerService<Article> jsonManagerService;
+    @Value("${articles.path}")
```
add an empty line before annotations (also check in other classes)

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/analyzer/AnalyzerService.java:
```diff
@@ -0,0 +1,102 @@
+
+@Service
+public class AnalyzerService {
+    List<String> analyzeResult = new ArrayList<>();
```
don't use global fields for mutable values

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/repository/ArticleRepository.java:
```diff
@@ -0,0 +1,25 @@
+@Repository
+public class ArticleRepository {
+    @Autowired
+    JsonManagerService<Article> jsonManagerService;
```
don't forget access modifier (private)

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/repository/CategoryRepository.java:
```diff
@@ -0,0 +1,26 @@
+@Repository
+public class CategoryRepository {
+    @Autowired
+    JsonManagerService<Categories> jsonManagerService;
```
same as above

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/repository/UserRepository.java:
```diff
@@ -0,0 +1,26 @@
+@Repository
+public class UserRepository {
+    @Autowired
+    JsonManagerService<User> jsonManagerService;
```
same as above

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+import java.util.List;
+import java.util.Set;
+
+@Service
```
why getter is needed?

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+    CategoryRepository categoryRepository;
+    @Autowired
+    AnalyzerService analyzerService;
+
```
use constructor injection and don't forget about encapsulation

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+        analyzerService.analyze(articles);
+    }
+
+    public void saveCategoriesToJSONFileAsObject(String fileName) {
+        categoryRepository.saveCategoriesToJSONFileAsObject(analyzerService.getCategories(), fileName);
```
this method is not used, should be removed

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/jsonmanager/JsonManagerService.java:
```diff
@@ -0,0 +1,57 @@
+public class JsonManagerService<T> {
+
+    @Autowired
+    private Gson gson;
```
constructor injection

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/jsonmanager/JsonManagerService.java:
```diff
@@ -0,0 +1,57 @@
+    @Autowired
+    private Gson gson;
+
+    private Set<T> data;
```
remove this global field

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/jsonmanager/JsonManagerService.java:
```diff
@@ -0,0 +1,57 @@
+        try {
+            FileWriter writer = new FileWriter(String.format("%s%s.json", path, fileName));
+            writer.write(json);
+            writer.close();
```
use try with resources

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/resources/application.properties:
```diff
@@ -0,0 +1,4 @@
+excluded.words=is,are,can,do,be,have,has,had,was,will,would,should,we,a,an,the,and,or,also,as,all,of,this,these,there,no,if,in,you
+users.path=article-analyzer/src/main/resources/models/users.json
+articles.path=article-analyzer/src/main/resources/models/articles.json
+categories.path=article-analyzer/src/main/resources/categories/
```
this does not work for me `Caused by: java.io.FileNotFoundException: article-analyzer\src\main\resources\categories\categories.json (The system cannot find the path specified)`

vitaliiStefanchakInventorsoft replied:
it reads articles, but cannot save the file (try to remove the one on your machine and run again)

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+        });
+    }
+
+    public List<String> getCategories() {
```
this method is needless, to granular, just use what it inside directly

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+
+    public void analyze() {
+        Set<Article> articles = articleRepository.readAllArticles();
+        analyzerService.analyze(articles);
```
you can call here saving categories and notifying users (these two methods bellow) as well, then the logic from the main method will not be needed

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+    public void notifyUsers(ConfigurableApplicationContext applicationContext) {
+        Set<User> users = userRepository.readAllUsers();
+        users.forEach(user -> {
+            notificationService = applicationContext.getBean(String.valueOf(user.getNotificationType()), NotificationService.class);
```
interesting idea, 
applicationContext maybe injected into this class as well as any other bean, so don't pass it as parameter

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/ArticleService.java:
```diff
@@ -0,0 +1,56 @@
+        Set<User> users = userRepository.readAllUsers();
+        users.forEach(user -> {
+            notificationService = applicationContext.getBean(String.valueOf(user.getNotificationType()), NotificationService.class);
+            notificationService.notifyUsers(user, getCategories());
```
also this global notificationService method should be local

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/analyzer/AnalyzerService.java:
```diff
@@ -0,0 +1,102 @@
+
+    }
+
+    public List<String> getCategories() {
```
this should be removed, the bean name should be stateless, and it should not store any state (data), the result should returned immediately in `analyze` method

vitaliiStefanchakInventorsoft commented on article-analyzer/src/main/java/co/inventorsoft/academy/articleanalyzer/service/analyzer/AnalyzerService.java:
```diff
@@ -0,0 +1,102 @@
+        return "\\b(" + escapedWords + ")\\b|\\W+";
+    }
+
+    public void analyze(Set<Article> articles) {
```
just a small comment here (can be avoided): the whole logic is correct, but it can be done in a more concise way.

vitaliiStefanchakInventorsoft reviewed (commented):
in general, the implementation looks good,
left the majority of comments about global fields, encapsulation and injection type

volyamnyi committed:
Refactoring according to the code review was done.

volyamnyi committed:
.gitkeep was added to the 'categories' folder.

volyamnyi committed:
Merge branch 'yamnyi-article-analyzer'

volyamnyi committed:
Beans from ApplicationConfig.java were moved to the appropriate classes.

volyamnyi committed:
Merge branch 'yamnyi-article-analyzer'

volyamnyi committed:
Calling the analyze method was moved to the main method;
.gitkeep was removed from the categories folder;
categories.json was added to the categories folder;

vitaliiStefanchakInventorsoft commented:
![ConsiderItDoneItsDoneGIF](https://github.com/vitaliiStefanchakInventorsoft/spring-lecture/assets/103140969/cc9ea718-299b-4a81-9741-0128bfc8ce1f)

volyamnyi commented:
Thanks a lot😁

нд, 15 жовт. 2023 р. о 13:28 Vitalii ***@***.***> пише:

> [image: ConsiderItDoneItsDoneGIF]
> <https://user-images.githubusercontent.com/103140969/275319823-cc9ea718-299b-4a81-9741-0128bfc8ce1f.gif>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/vitaliiStefanchakInventorsoft/spring-lecture/pull/4#issuecomment-1763347117>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AVLB4H3MWUMIVSAJN4F37ZTX7O3GTAVCNFSM6AAAAAA57A6ZTOVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTONRTGM2DOMJRG4>
> .
> You are receiving this because you were assigned.Message ID:
> ***@***.***
> com>
>
